openghg/objectstore/metastore/_metastore.py

```diff
@@ -0,0 +1,242 @@
+"""
+This module defines an interface for metastores.
+
+Metastores store metadata that is used to search an
+object store for data.
+
+The documentation in this module refers to metadata entries stored
+in the metastore as "records". These correspond to "documents" in
+TinyDB and Mongodb.
+
+A MetaStore managed by an ObjectStore object will store
+UUIDs, which are used to locate data stored in the object store.
+
+UUIDs are managed by ObjectStore objects, so they are not part of
+the MetaStore interface.
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Union
+
+import tinydb
+from tinydb.operations import delete as tinydb_delete
+
+from openghg.types import MetastoreError
+
+MetaData = Dict[str, Any]
+QueryResults = List[Dict[str, Any]]  # ...to avoid clashes with `SearchResults` object
+Bucket = str
+
+
+class MetaStore(ABC):
+    """Interface for MetaStore.
+
+    All classes implementing this interface must provide the following methods:
+        search: to search for records in the metastore matching given key-value pairs
+        add: to add metadata to a new record in the metastore
+        delete: to delete records matching the given key-value pairs
+        update: update the metadata of a record matching the `where` key-value pairs;
+            keys may be added or updated (using `to_update`) or deleted (using `to_delete`)
+    """
+
+    @abstractmethod
+    def search(self, search_terms: Optional[MetaData] = None) -> QueryResults:
+        """Search for data using a dictionary of search terms.
+
+        Args:
+            search_terms: dictionary of key-value pairs to search by.
+
+        Returns:
+            list of records in the metastore matching the given search terms.
+            If search_terms is None, return all records.
+        """
+        pass
+
+    @abstractmethod
+    def add(self, metadata: MetaData) -> None:
```

Comment:
In general, the meaning of *add* is a bit loose in database design (not fully agreed). For the TinyDB implementation below the insert function has been used which does have a specific meaning - "add this, but error if it already exists" (which contrasts with other *add* type options "update" which means "replace the existing thing, and error it if *doesn't* exist" and "upsert" which is a combination which means "update this thing if it exists, or add it if it doesn't") - [TinyDB's definitions](https://tinydb.readthedocs.io/en/latest/usage.html) and this is also consistent with SQL.

This is a long-winded way of saying, I think the name `insert` is better here, but would also require an explicit comment in the docstring saying "it will error it it already exists". That way, when adding different subclasses in the future, the interface is explicitly defined.

Reply:
Sure sounds good. I can't require the exception in the interface but I will mention it in the docstring.

Reply:
I don't think TinyDB's insert will throw an error unless the Document you're adding already has a document id, and this document id is already in the database. The analogous behaviour for the object store is to throw an error if we try to add/insert metadata that already has a UUID, and this UUID is already used.

So, I'll make the name change, but I won't add the insert/update error to the metastore at this point.

(Somewhat related to this: I moved some error handling to the ABC, since it only depends on `search`.)

One option is to make a mixin if we want strict inserts:
`
class StrictInsertMixin:
    def insert(self, metadata):
        if len(super().search(metadata)) > 1:
            raise MetastoreError
        super().insert(metadata)
`

Reply:
I changed the name of the method to `insert`. I think if we want to raise errors on inserting metadata that matches existing metadata (ignoring document IDs), the mixin is a decent solution.

At some point, we will need to create `MetaStore`s based on configs set in a bucket, so no one will manually have to type the name `StrictTinyDBMetaStore`, they can just call `open_metastore`.

Reply:
Ok so with the current design using insert through TinyDB multiple times with the same UUID would not throw an exception because UUID is not identified as the Document ID? And it looks like we can't set the Document ID to be the UUID because this needs [to be an integer](https://tinydb.readthedocs.io/en/latest/api.html#tinydb.table.Document).

I feel like the Mixin solution may be overly complex - it would be fine to make it clear what we expect the child class to do within the documentation. We could add this check to the TinyDB insert method to make sure this is enforced.

Reply:
Yes with the current design, you can insert the same data as many times as you like (each instance would have a different document ID, created by TinyDB) unless you specify the document ID as one that already exists.

The metastore isn't responsible for UUIDs at all, that's done currently by `BaseStore` and will be done by `ObjectStore` in the future. I think it makes more sense for UUIDs to be handled at the "object store" level, since it is information needed by both the metastore and datasources.

The metastore ABC is just meant to be an abstraction over the database, and the simpler the better, provided we can do everything we need with the operations it provides.

I can add a check to require that the particular set of key-value pairs is unique when data is inserted. To be consistent, we would have to add checks to make sure that updating metadata doesn't cause it to duplicate any existing data.

I added the checks to update and delete because I think accidentally deleting a lot of data would be annoying, and I put a requirement that only one item is updated at a time, since the logic for updating multiple items is significantly more complex, and we could just use a for loop to do it if we need to. 

At the object store level, we can manage which keys are immutable, so it won't be possible to update the uuid of an item, and so updating can never duplicate another record. 

TLDR: ensuring metastore entries are unique will be easier to do within the object store.

I'll make a draft PR with the `ObjectStore` class so you can see what I mean.

Reply:
This is the branch with the `ObjectStore` class: iss766_part_4_objectstore_with_ABC_datasource